app/services/applied_coupons/create_service.rb

```diff
@@ -96,6 +97,16 @@ def reusable_coupon?
       customer.applied_coupons.where(coupon_id: coupon.id).none?
     end
 
+    def plan_limitation_overlapping?
+      return false unless coupon.limited_plans?
+
+      related_plans = coupon.coupon_plans.pluck(:plan_id)
+
+      active_coupons_ids = customer.applied_coupons.active.pluck(:coupon_id)
+
+      CouponPlan.where(coupon_id: active_coupons_ids, plan_id: related_plans).exists?
```

Comment:
I think you can achieve the same think with one query instead of 3: (To be validated ;) )
```ruby
customer.applied_coupons
  .active
  .joins(coupon: :coupon_plans)
  .where(coupon_plans: { plan_id: coupon.coupon_plans.select(:plan_id) })
  .exist?
```